src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs

```diff
@@ -0,0 +1,1451 @@
+        /// </exception>
+        protected abstract bool VerifyDataCore(ReadOnlySpan<byte> data, ReadOnlySpan<byte> context, ReadOnlySpan<byte> signature);
+
+        /// <summary>
+        ///   When overridden in a derived class, exports the FIPS 204 public key to the specified buffer.
+        /// </summary>
+        /// <param name="destination">
+        ///   The buffer to receive the public key.
+        /// </param>
+        protected abstract void ExportMLDsaPublicKeyCore(Span<byte> destination);
+
+        /// <summary>
+        ///   When overridden in a derived class, exports the FIPS 204 secret key to the specified buffer.
+        /// </summary>
+        /// <param name="destination">
+        ///   The buffer to receive the secret key.
+        /// </param>
+        protected abstract void ExportMLDsaSecretKeyCore(Span<byte> destination);
+
+        /// <summary>
+        ///   When overridden in a derived class, exports the private seed to the specified buffer.
+        /// </summary>
+        /// <param name="destination">
+        ///   The buffer to receive the private seed.
+        /// </param>
+        protected abstract void ExportMLDsaPrivateSeedCore(Span<byte> destination);
+
+        private AsnWriter ExportSubjectPublicKeyInfoCore()
+        {
+            ThrowIfDisposed();
+
+            byte[] rented = CryptoPool.Rent(_parameterSetInfo.PublicKeySizeInBytes);
+
+            try
+            {
+                Span<byte> keySpan = rented.AsSpan(0, _parameterSetInfo.PublicKeySizeInBytes);
+                ExportMLDsaPublicKey(keySpan);
+
+                AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+
+                using (writer.PushSequence())
+                {
+                    using (writer.PushSequence())
+                    {
+                        writer.WriteObjectIdentifier(_parameterSetInfo.Oid);
+                    }
+
+                    writer.WriteBitString(keySpan);
+                }
+
+                return writer;
+            }
+            finally
+            {
+                // Public key doesn't need to be cleared
+                CryptoPool.Return(rented, 0);
+            }
+        }
+
+        private AsnWriter ExportEncryptedPkcs8PrivateKeyCore(ReadOnlySpan<byte> passwordBytes, PbeParameters pbeParameters)
+        {
+            ThrowIfDisposed();
+
+            // TODO: Determine a more appropriate maximum size once the format is actually known.
+            int initialSize = _parameterSetInfo.SecretKeySizeInBytes * 2;
+            // The buffer is only being passed out as a span, so the derived type can't meaningfully
+            // hold on to it without being malicious.
+            byte[] rented = CryptoPool.Rent(initialSize);
+            int written;
+
+            while (!TryExportPkcs8PrivateKey(rented, out written))
+            {
+                CryptoPool.Return(rented, 0);
+                rented = CryptoPool.Rent(rented.Length * 2);
```

Comment:
I realize this is safe, but seeing `rented` used after it's returned makes me wince.

Reply:
I had the same feeling when I wrote it... so now it doesn't do that any more.

(Every now and then I try to write smaller code, see? :smile:)